crates/resolver/src/name_server/name_server_pool.rs

```diff
@@ -217,6 +217,15 @@ where
                         debug!("truncated response received, retrying over TCP");
                         Ok(response)
                     }
+                    Err(e)
+                        if opts.try_tcp_on_io_error_only
+                            && opts.try_tcp_on_error
+                            && !e.is_no_connections()
+                            && !e.is_io() =>
+                    {
+                        debug!("not an I/O error, not retrying over TCP: {}", e);
+                        return Err(e);
+                    }
                     Err(e) if opts.try_tcp_on_error || e.is_no_connections() || e.is_io() => {
```

Comment:
Ok, this seems like I got this wrong a few months ago in #2215.

So here's my opinion. `try_tcp_on_error` should probably only be associated with io errors. the e.is_no_connections is a special case for no UDP connections and only TCP is available. So we always want to use TCP in that situation.

I think we can avoid the new config param if we change this to `if (opts.try_tcp_on_error && e.is_io()) || e.is_no_connections()`